### DISCUSSION TITLE
new feature breaks backwards compatiability

### DISCUSSION THREAD

**Discussion by davidhadas** [minio/minio#12673] [Q&A] [answered]
## Expected Behavior
Backwards compatibility

## Current Behavior
Running minio from latest, minio stopped working due to https://github.com/minio/minio/commit/cdeccb5510f25d0019682f978c2c0aa02a5e12bf
Reason being 
1. changes in the default externalized behaviour 
2. existing config led to minio not working and writing to screen the error from this code:
        "WARNING: Console endpoint is listening on a dynamic port (%s), please use --console-address \":PORT\" to choose a static port.",

## Possible Solution
Keeping backwards compatibility should be mandatory when ever feasible.
Adding a config for a separate console port should not change backwards compatibility

## Steps to Reproduce (for bugs)
Most likely you can see this when starting minio in Kube with this spec (have not tried it..)
spec:
      containers:
          ports:
            - containerPort: 9000
              protocol: TCP
          image: 'minio/minio'
          args:
            - server
            - /storage

Instead, I was using a more elaborate spec:
spec:
      containers:
        - resources: {}
          name: minio
          env:
            - name: MINIO_ROOT_USER
              value: minio
            - name: MINIO_ROOT_PASSWORD
              value: M1N10OSpss
            - name: MINIO_NOTIFY_KAFKA_ENABLE_HOOK
              value: 'on'
            - name: MINIO_NOTIFY_KAFKA_BROKERS_HOOK
              value: 'kafka-service:9092'
            - name: MINIO_NOTIFY_KAFKA_TOPIC_HOOK
              value: mytest
          ports:
            - containerPort: 9000
              protocol: TCP
          imagePullPolicy: Always
          image: 'minio/minio'
          args:
            - server
            - /storage

## Context
I encountered this issue as I restarted my pod and since I am using a policy to always pull the image

## Regression

No

## Your Environment
Running Minio on Openshift with today's minio from docket hub

**Comment by harshavardhana** [comment, marked as answer]
@davidhadas there is no backward compatibility in UI - it is a project-wide decision. We are not planning to change back anything here. 

Documentation is updated on what is done and release notes are updated on what to expect. If you do not wish to upgrade don't do it. If you upgrade keep in mind things to consider.

https://github.com/minio/minio#test-using-minio-console

**Comment by davidhadas** [comment]
@harshavardhana 
1. Classifying this as a  UI issue is wrong. UI backwards compatibility is a minor issue - this is a backwards comaptiability issue of the basic service config -  and therefore users will be affected..!
2. When projects have minor and major releases, it is expected that major releases will include non backwards compatiability  but minor changes will not include such issues. 
3. Asking users to freeze to a version means they do not get security updates  - so this is not a working solutions for users. 

I am not making any statements against having this feature
All I am saying  is that  when a user have a working setup of minio with some existing config  and suddenly it stops working since the minio team decided to "add a feature" - this is plain bad for the project and its users

**Comment by harshavardhana** [comment]
> All I am saying is that when a user have a working setup of minio with some existing config and suddenly it stops working since the minio team decided to "add a feature" - this is plain bad for the project and its users

I don't think anything of this is broken here all the features work as is @davidhadas - I tried your spec it all works fine locally for me. 

You are not saying anything here other than not showing what the error is.

**Comment by davidhadas** [comment]
Q: You are not saying anything here other than not showing what the error is.

A: As indicated above, I received the error that was triggered by the line of code: 
"WARNING: Console endpoint is listening on a dynamic port (%s), please use --console-address ":PORT" to choose a static port." 
And minio services did not start.

The same config with a version from some days earlier were working ok.

**Reply by harshavardhana** [reply]
That is a warning not an error

**Reply by davidhadas** [reply]
I think you are missing the point.

A DevOp deploys Minio and many other dependencies as part of his workload using Kubernetes for example, let's say that the DevOps chose port 9000 and exposes it as his API and console   - in line with Minio documentation. 
The entire workload is tested and working for long time. 
The workload users are happy. 

One day Minio team decided to "add a feature" and the next time the image is taken from Docker Hub (e.g. when the pod reboots) suddenly the workload stop working. The users complain that the console is not available. The reason is that the previously exposed console on port 9000 is no longer working although the DevOp did not change his configuration. This means that the DevOp previous working setup is suddenly not working due to some decision made by the Minio Team, although the DevOp did not change anything on his side. 

This is not the way reliable packages work.

**Reply by davidhadas** [reply]
What you should do instead is allow the console to be the same as the API by default and document that it is no longer recommended (if that is the case). You can say that having the two the same is deprecated, you can issue warnings, but you cant move to a dynamic arbitrary port for console since this changes the exposed behavior in a major way.

**Reply by harshavardhana** [reply]
@davidhadas  It is mentioned in the docs here, release notes and also other elaborate documentation provided to ensure that you read these things before doing any willy nilly upgrades.

- https://github.com/minio/minio#test-using-minio-console 
- https://github.com/minio/minio/releases/tag/RELEASE.2021-07-08T01-15-01Z
 
Beyond this, I think I shall close this topic.

**Comment by xoxys** [comment]
How does the Prometheus and health endpoint work with the latest version? `/minio/v2/metrics/cluster` and `/minio/health/live` does not work for me on the console port or the minio port.

**Reply by xoxys** [reply]
Got it to work, my fault.
